python.d/ntpd.chart.py

```diff
@@ -0,0 +1,385 @@
+
+    def rebuild_request(self, new_idx):
+        self.id = new_idx
+        self.request = self.get_header(self.id)
+
+
+class Service(SocketService):
+    def __init__(self, configuration=None, name=None):
+        SocketService.__init__(self, configuration=configuration, name=name)
+        self.order = list(ORDER)
+        self.definitions = dict(CHARTS)
+        self.definitions.update(PEER_CHARTS)
+
+        self.port = 'ntp'
+        self.dgram_socket = True
+        self.system = System()
+        self.peers = dict()
+        self.request = str()
+        self.retries = 0
+        self.all_peers = self.configuration.get('all_peers', False)
+
+        peer_filter = self.configuration.get('peer_filter', r'127\..*')
+        try:
+            self.peer_filter = re.compile(r'^((0\.0\.0\.0)|({0}))$'.format(peer_filter))
+        except re.error as error:
+            self.error('Compile pattern error : {0}'.format(error))
+            self.peer_filter = None
+
+    def check(self):
+        """
+        Checks if we can get valid systemvars.
+        If not, returns None to disable module.
+        """
+
+        if not self.peer_filter:
+            return None
+
+        self._parse_config()
+
+        self.request = self.system.request
+        raw_systemvars = self._get_raw_data()
+
+        if not self.system.get_data(raw_systemvars):
+            return None
+        return True
+
+    def get_data(self):
+        """
+        Gets systemvars data on each update.
+        Gets peervars data for all peers on each update.
+        """
+        data = dict()
+
+        self.request = self.system.request
+        raw = self._get_raw_data()
+        if not raw:
+            return None
+
+        data.update(self.system.get_data(raw))
+
+        if not self.all_peers:
+            return data
+
+        # TODO: run self.find_new_peers() every N runs?
+        if not self.peers or self.retries > 8:
+            self.find_new_peers()
+        else:
+            for peer in self.peers.values():
+                self.request = peer.request
+                peer_data = peer.get_data(self._get_raw_data())
+                if peer_data:
+                    data.update(peer_data)
+                else:
+                    self.retries += 1
+
+        return data
+
+    def find_new_peers(self):
+        new_peers = dict((p.real_name, p) for p in self.get_peers())
+        if new_peers:
+
+            peers_to_remove = set(self.peers) - set(new_peers)
+            peers_to_update = set(self.peers) & set(new_peers)
+            peers_to_add = set(new_peers) - set(self.peers)
+
+            for peer_name in peers_to_remove:
+                self.hide_old_peer_from_charts(self.peers[peer_name])
+                del self.peers[peer_name]
+
+            for peer_name in peers_to_update:
+                if new_peers[peer_name].id != self.peers[peer_name].id:
+                    self.peers[peer_name].rebuild_request(new_peers[peer_name].id)
```

Comment:
looking through the code i found that
`peers_to_update` and `for peer_name in peers_to_update:` not needed.

```python
self.peers.update(new_peers)
```

do the job

(`Peer.rebuild_request` should also be deleted)